auth/couchdb_session_authenticator_test.go

```diff
@@ -0,0 +1,281 @@
+			{"", "password"},
+			{"{invalid-user}", "password"},
+			{"user", ""},
+			{"user", "{invalid-password}"},
+		}
+
+		for _, tt := range errortests {
+			_, err := NewCouchDbSessionAuthenticator(tt.user, tt.password)
+			Expect(err).To(HaveOccurred())
+
+			_, err = NewCouchDbSessionAuthenticatorFromMap(map[string]string{
+				"USERNAME": tt.user,
+				"PASSWORD": tt.password,
+			})
+			Expect(err).To(HaveOccurred())
+		}
+	})
+
+	Context("with standard test server", func() {
+		var (
+			server  *httptest.Server
+			request *http.Request
+			auth    *CouchDbSessionAuthenticator
+			err     error
+		)
+
+		BeforeEach(func() {
+			callNumber := 0
+			server = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				callNumber++
+				cookie := &http.Cookie{
+					Name:  "AuthSession",
+					Value: fmt.Sprintf("fakefake-%d", callNumber),
+					Expires: time.Now().
+						Add(24*time.Hour - time.Duration(callNumber)*time.Minute),
+				}
+				http.SetCookie(w, cookie)
+				w.WriteHeader(http.StatusOK)
+			}))
+
+			builder, err := core.NewRequestBuilder(core.GET).
+				ResolveRequestURL(server.URL, "/db", nil)
+			Expect(err).To(BeNil())
+
+			ctx := context.WithValue(context.Background(), contextKey("key"), "abcdefgh")
+			request, err = builder.
+				AddHeader("X-Req-Id", "abcdefgh").
+				WithContext(ctx).
+				Build()
+			Expect(err).To(BeNil())
+			Expect(request).ToNot(BeNil())
+
+			auth, err = NewCouchDbSessionAuthenticator("user", "pass")
+			Expect(err).To(BeNil())
+			Expect(auth).ToNot(BeNil())
+			Expect(auth.AuthenticationType()).To(Equal(AUTHTYPE_COUCHDB_SESSION))
+		})
+
+		AfterEach(func() {
+			server.Close()
+		})
+
+		It("Test setting URL, Headers and Context on Authenticator", func() {
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			Expect(auth.url).To(Equal(server.URL))
+			Expect(auth.header).To(HaveKeyWithValue("X-Req-Id", []string{"abcdefgh"}))
+			Expect(auth.ctx.Value(contextKey("key"))).To(Equal("abcdefgh"))
+		})
+
+		It("Test setting custom http client on Authenticator", func() {
+			// set http.Client with custom timeout
+			auth.Client = &http.Client{Timeout: time.Second}
+
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			Expect(auth.Client.Timeout).To(Equal(time.Second))
+		})
+
+		It("Test authentication re-request after expiration", func(done Done) {
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			// Fetch a cookie from the cache to confirm it's present
+			// and prepare cache state for the next test
+			cookie, err := auth.getCookie()
+			Expect(err).To(BeNil())
+			Expect(cookie).ToNot(BeNil())
+			Expect(cookie.Name).To(Equal("AuthSession"))
+			Expect(cookie.Value).To(Equal("fakefake-1"))
+			Expect(cookie.Expires).ToNot(BeNil())
+
+			// Fetch cookie again to verify cache is working as expected,
+			// we are getting old cookie and refresh process's not triggered
+			oldRefresh := auth.session.refreshTime
```

Comment:
1. This check should be made with the first cache fetch, too.
2. I cannot see the point of this second cache fetch.

Reply:
1. What check? I'm storing old refresh time here so further on I can confirm the refresh time will advance.
1. Right, I see you are struggling with this test, consider number of the comments you left. I goes like this:

- Do `auth.Authenticate` to get cookie and its times into the cache.
- Do first cache fetch and confirm `Authenticate` worked and we have out cookie and its times.
- Store refresh time, do _second_ cache fetch and verify we still have old cookie and refresh time, to confirm that we didn't accidentally triggered offline refresh with the _previous_ cache fetch.
- Move expiration time, do three cache fetches in parallel and verify that we've got one and the same new cookie and new refresh time, so our mutex works and we only starting one offline cache refresh.

In other words second fetch here is to verify that we are not accidentally triggering offline cache refresh on every fetch. Hope this helps.